src/platform_impl/windows/keyboard.rs

```diff
@@ -0,0 +1,809 @@
+    platform_impl::platform::{
+        event_loop::ProcResult,
+        keyboard_layout::{get_or_insert_str, Layout, LayoutCache, WindowsModifiers, LAYOUT_CACHE},
+        KeyEventExtra,
+    },
+};
+
+pub fn is_msg_keyboard_related(msg: u32) -> bool {
+    use winuser::{WM_KEYFIRST, WM_KEYLAST, WM_KILLFOCUS, WM_SETFOCUS};
+    let is_keyboard_msg = WM_KEYFIRST <= msg && msg <= WM_KEYLAST;
+
+    is_keyboard_msg || msg == WM_SETFOCUS || msg == WM_KILLFOCUS
+}
+
+pub type ExScancode = u16;
+
+pub struct MessageAsKeyEvent {
+    pub event: KeyEvent,
+    pub is_synthetic: bool,
+}
+
+/// Stores information required to make `KeyEvent`s.
+///
+/// A single Winit `KeyEvent` contains information which the Windows API passes to the application
+/// in multiple window messages. In other words: a Winit `KeyEvent` cannot be built from a single
+/// window message. Therefore, this type keeps track of certain information from previous events so
+/// that a `KeyEvent` can be constructed when the last event related to a keypress is received.
+///
+/// `PeekMessage` is sometimes used to determine whether the next window message still belongs to the
+/// current keypress. If it doesn't and the current state represents a key event waiting to be
+/// dispatched, then said event is considered complete and is dispatched.
+///
+/// The sequence of window messages for a key press event is the following:
+/// - Exactly one WM_KEYDOWN / WM_SYSKEYDOWN
+/// - Zero or one WM_DEADCHAR / WM_SYSDEADCHAR
+/// - Zero or more WM_CHAR / WM_SYSCHAR. These messages each come with a UTF-16 code unit which when
+///   put together in the sequence they arrived in, forms the text which is the result of pressing the
+///   key.
+///
+/// Key release messages are a bit different due to the fact that they don't contribute to
+/// text input. The "sequence" only consists of one WM_KEYUP / WM_SYSKEYUP event.
+pub struct KeyEventBuilder {
+    event_info: Option<PartialKeyEventInfo>,
+
+    /// The keyup event needs to call `ToUnicode` to determine what the text produced by the
+    /// key with all modifiers except CTRL (the `logical_key`) is.
+    ///
+    /// But `ToUnicode` without the non-modifying flag (see `key_labels`), resets the dead key
+    /// state which would be incorrect during every keyup event. Therefore this variable is used
+    /// to determine whether the last keydown event produced a dead key.
+    ///
+    /// Note that this variable is not always correct because it does
+    /// not track key presses outside of this window. However, the ONLY situation where this
+    /// doesn't work as intended is when the user presses a dead key outside of this window, and
+    /// switches to this window BEFORE releasing it then releases the dead key. In this case
+    /// the `ToUnicode` function will be called, incorrectly clearing the dead key state. Having
+    /// an inccorect behaviour only in this case seems acceptable.
+    prev_down_was_dead: bool,
```

Comment:
reported by CI as never read

Reply:
Oh this was apparently left here from an earlier state when a completely different approach was used.